# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/jit/arm/LIR-arm.h
# Commit: 83a3ed07220d
# Full Hash: 83a3ed07220dfe4c8a2d16a2c48cfa8d49635362
# Author: Dmitry Bezhetskov <dbezhetskov@igalia.com>
# Date: 2020-09-14 09:51:01
# Regressor Bug: 1639153
# File Overlap Count: 1
# Description:
#   Bug 1639153 - Part 6.3: Establish dependency from tls for arm callWithABI div/mod i64. r=lth
#   
#   To be able to call c++ runtime via Builtin thunk we need to set up WasmTlsReg.
#   In this patch I create dependencies from MIR level to Codegen to be sure that WasmTlsReg is alive
#   when we call runtime in div/mod i64 for arm.
# ==============================================================================

diff -r ca66708049b8 -r 83a3ed07220d js/src/jit/arm/LIR-arm.h
--- a/js/src/jit/arm/LIR-arm.h	Mon Sep 14 02:54:53 2020 +0000
+++ b/js/src/jit/arm/LIR-arm.h	Mon Sep 14 04:00:18 2020 +0000
@@ -95,80 +95,86 @@
 };
 
 class LDivOrModI64
-    : public LCallInstructionHelper<INT64_PIECES, INT64_PIECES * 2, 0> {
+    : public LCallInstructionHelper<INT64_PIECES, INT64_PIECES * 2 + 1, 0> {
  public:
   LIR_HEADER(DivOrModI64)
 
   static const size_t Lhs = 0;
   static const size_t Rhs = INT64_PIECES;
+  static const size_t Tls = 2 * INT64_PIECES;
 
-  LDivOrModI64(const LInt64Allocation& lhs, const LInt64Allocation& rhs)
+  LDivOrModI64(const LInt64Allocation& lhs, const LInt64Allocation& rhs,
+               const LAllocation& tls)
       : LCallInstructionHelper(classOpcode) {
     setInt64Operand(Lhs, lhs);
     setInt64Operand(Rhs, rhs);
+    setOperand(Tls, tls);
   }
 
-  MBinaryArithInstruction* mir() const {
-    MOZ_ASSERT(mir_->isDiv() || mir_->isMod());
-    return static_cast<MBinaryArithInstruction*>(mir_);
+  MDefinition* mir() const {
+    MOZ_ASSERT(mir_->isWasmBuiltinDivI64() || mir_->isWasmBuiltinModI64());
+    return mir_;
   }
   bool canBeDivideByZero() const {
-    if (mir_->isMod()) {
-      return mir_->toMod()->canBeDivideByZero();
+    if (mir_->isWasmBuiltinModI64()) {
+      return mir_->toWasmBuiltinModI64()->canBeDivideByZero();
     }
-    return mir_->toDiv()->canBeDivideByZero();
+    return mir_->toWasmBuiltinDivI64()->canBeDivideByZero();
   }
   bool canBeNegativeOverflow() const {
-    if (mir_->isMod()) {
-      return mir_->toMod()->canBeNegativeDividend();
+    if (mir_->isWasmBuiltinModI64()) {
+      return mir_->toWasmBuiltinModI64()->canBeNegativeDividend();
     }
-    return mir_->toDiv()->canBeNegativeOverflow();
+    return mir_->toWasmBuiltinDivI64()->canBeNegativeOverflow();
   }
   wasm::BytecodeOffset bytecodeOffset() const {
-    MOZ_ASSERT(mir_->isDiv() || mir_->isMod());
-    if (mir_->isMod()) {
-      return mir_->toMod()->bytecodeOffset();
+    MOZ_ASSERT(mir_->isWasmBuiltinDivI64() || mir_->isWasmBuiltinModI64());
+    if (mir_->isWasmBuiltinModI64()) {
+      return mir_->toWasmBuiltinModI64()->bytecodeOffset();
     }
-    return mir_->toDiv()->bytecodeOffset();
+    return mir_->toWasmBuiltinDivI64()->bytecodeOffset();
   }
 };
 
 class LUDivOrModI64
-    : public LCallInstructionHelper<INT64_PIECES, INT64_PIECES * 2, 0> {
+    : public LCallInstructionHelper<INT64_PIECES, INT64_PIECES * 2 + 1, 0> {
  public:
   LIR_HEADER(UDivOrModI64)
 
   static const size_t Lhs = 0;
   static const size_t Rhs = INT64_PIECES;
+  static const size_t Tls = 2 * INT64_PIECES;
 
-  LUDivOrModI64(const LInt64Allocation& lhs, const LInt64Allocation& rhs)
+  LUDivOrModI64(const LInt64Allocation& lhs, const LInt64Allocation& rhs,
+                const LAllocation& tls)
       : LCallInstructionHelper(classOpcode) {
     setInt64Operand(Lhs, lhs);
     setInt64Operand(Rhs, rhs);
+    setOperand(Tls, tls);
   }
 
-  MBinaryArithInstruction* mir() const {
-    MOZ_ASSERT(mir_->isDiv() || mir_->isMod());
-    return static_cast<MBinaryArithInstruction*>(mir_);
+  MDefinition* mir() const {
+    MOZ_ASSERT(mir_->isWasmBuiltinDivI64() || mir_->isWasmBuiltinModI64());
+    return mir_;
   }
   bool canBeDivideByZero() const {
-    if (mir_->isMod()) {
-      return mir_->toMod()->canBeDivideByZero();
+    if (mir_->isWasmBuiltinModI64()) {
+      return mir_->toWasmBuiltinModI64()->canBeDivideByZero();
     }
-    return mir_->toDiv()->canBeDivideByZero();
+    return mir_->toWasmBuiltinDivI64()->canBeDivideByZero();
   }
   bool canBeNegativeOverflow() const {
-    if (mir_->isMod()) {
-      return mir_->toMod()->canBeNegativeDividend();
+    if (mir_->isWasmBuiltinModI64()) {
+      return mir_->toWasmBuiltinModI64()->canBeNegativeDividend();
     }
-    return mir_->toDiv()->canBeNegativeOverflow();
+    return mir_->toWasmBuiltinDivI64()->canBeNegativeOverflow();
   }
   wasm::BytecodeOffset bytecodeOffset() const {
-    MOZ_ASSERT(mir_->isDiv() || mir_->isMod());
-    if (mir_->isMod()) {
-      return mir_->toMod()->bytecodeOffset();
+    MOZ_ASSERT(mir_->isWasmBuiltinDivI64() || mir_->isWasmBuiltinModI64());
+    if (mir_->isWasmBuiltinModI64()) {
+      return mir_->toWasmBuiltinModI64()->bytecodeOffset();
     }
-    return mir_->toDiv()->bytecodeOffset();
+    return mir_->toWasmBuiltinDivI64()->bytecodeOffset();
   }
 };
 